[GSoC] [WIP] NeuroEvolution of Augmenting Topologies

favre49 opened the pull request:
This is a WIP PR for an implementation of NEAT and its tests. I'll be adding progress updates in the comments.

favre49 commented:
I've just added the test suite. The `BOOST_REQUIRE` lines are still placeholders, since I am not sure what to keep as the test conditions. I've also scaled down the variables in the Double Pole balancing cases, there are two things i wanted to clarify:
- I've scaled the velocity variables by dividing by 100, since we don't have bounds on them. Is this fine?
- If this is fine, should i do the same in all the test cases? I think that would be appropriate, but perhaps we should wait to see what kind of results the final implementation gives on unscaled variables? Let me know what you think.

zoq commented:
>  I think that would be appropriate, but perhaps we should wait to see what kind of results the final implementation gives on unscaled variables? Let me know what you think.

NEAT should be able to evolve a normalisation step if needed, but I agree it might be useful, to scale the input, perhaps we should add a flag that allows us to train on scaled or unscaled data.

favre49 commented:
Hmmm what exactly do you mean by that?

favre49 commented:
I've added implementations of the Gene classes. Some additional documentation and style fixing needs to be done, I'll do that soon.

zoq commented:
> Hmmm what exactly do you mean by that?
I don't think scaling is something we have to do, but I also think it could help to accelerate the training process. As you already said, let's keep it in for now and once we have something for testing we can check if it makes a difference.

zoq commented on src/mlpack/methods/neat/gene.hpp:
```diff
@@ -0,0 +1,101 @@
+/**
+ * Connection gene class.
+ */
+class ConnectionGene
```
Do you think you could draw the overall class connection infrastructure (what class depends on which, connection between classes) on a paper? It doesn't have to be fancy.

favre49 replied:
![IMG_20190601_193935015_HDR](https://user-images.githubusercontent.com/40389657/58749566-3c70b380-84a5-11e9-8e23-98b67eaf6be0.jpg)
Do you mean something like this?

zoq replied:
Yes, thanks that is really helpful. So I think you will start with the the implementation of the Genome? You already have the interface for the ConnectionGene. which will be part of the Genome that also holds a list of Genes? If you like I can sketch up a high level interface?

favre49 replied:
Yup, I'm working on the implementation of the Genome class right now. In the genome I am planning to maintain two lists of Connection Genes, where one will be an adjacency list to help with evaluating the phenomes and the other is a std::vector storing the genes sorted by innovation ID. As for the interface, besides what we discussed Input() and Output()) what else would the user need? The population would need to use a Mutate() and Crossover() function. I don't think there is much else.

manish7294 replied:
Maybe here instead of storing genes as a vector, you can just store pointers to genes sorted by innvoation ID.

favre49 replied:
What would be the benefit to this over storing the genes directly? I think if we use pointers, I would have to dynamically allocate the genes, which would really complicate things with having to concern ourselves with memory management. Is there something I'm missing?

manish7294 replied:
I intend to find a way to reduce the space usage here. But, you are right about the issue of memory management, I completely missed out this point. And if I recall correctly, we usually don't do pointers here in mlpack.

zoq replied:
You could use ``ptr_vector``, or I guess you can also use move semantics to avoid another copy, but at the end you have to store the object somewhere.

> As for the interface, besides what we discussed Input() and Output()) what else would the user need? 


I was thinking about a more in depth internal class interface, this class will hold this object, this class is called by another class using this interface, etc.

favre49 replied:
Oh, I understand. I can't think of much, besides making a class wrapping all the gene storage containers, which makes it easier to pass to the Phenome class? Maybe also a class that stores the "config" of the genome? Like the means for the weights and biases, input and output node counts, maximum complexity for pruning, etc. Once again we wouldn't have to use a lot of parameters to pass basic info to other methods and classes.

zoq replied:
Agreed, we should defiantly think about how we want to load a config, but don't think this has priority right now. If you already have a ``Gene`` class in mind, please feel free to commit something; makes it easier to discuss the interface anyway. Also, personally I would go with a simple container e.g. ``std::vector`` for now @manish7294 do you think that is reasonable?

manish7294 replied:
Sure, we can go with vector. And can always at later point of time, see what can be improved.

favre49 commented:
I have added the Genome class as well as the class for evaluation of acyclic networks. Style checks and some documentation still remain. I also have to prevent cycles from forming during mutation if the `isAcyclic` param is set to `true`.

How the acyclic network evaluation works is:
1. Assign depth to all of the nodes. Input and bias nodes are of course depth 0. In the case that multiple depths are found, we choose the maximum depth.
2. Based on these depths, nodes are split into layers.
3. Then, we activate the nodes layer by layer and take the output.

If there are any inefficiencies, please point them out. Another thing I will have to do is make some changes in the test suite, since the genomes now have the activation function as a template parameter.

zoq commented on src/mlpack/methods/neat/acyclic_net.hpp:
```diff
@@ -0,0 +1,74 @@
+   * @param outputNodeCount The number of output nodes.
+   * @param bias The bias.s
+   */
+  AcyclicNet(std::vector<int>& NodeGeneList,
```
I'm wondering if it would be more efficient to update the network instead of creating a new one for each evalution. let me know what you think.

favre49 replied:
I agree, but I was not able to think of an appropriate way to do this during crossovers, since the entire network structure undergoes a change, unlike mutation where I could update depths for a single node or connection. I'll think about it some more though.

zoq replied:
What we could do is to update the overall network once we do crossovers and update the it in the other case. But ideally we can do both without rebuilding the network, I'll think about it as well.

favre49 replied:
The thing is, mutation and crossover occur together during the reproduction step, so under the current system, we don't have to update for mutation either, since we aren't for crossover.

I can however remove the node gene list, since nodes can't be deleted, i just need to store the max node ID. I'm also working on the cyclic network evaluation right now.

zoq replied:
You are right thanks for the clarification.

zoq commented on src/mlpack/methods/neat/genome_impl.hpp:
```diff
@@ -0,0 +1,165 @@
+
+// Evaluates output based on input.
+template <class ActivationFunction>
+arma::vec Genome<ActivationFunction>::Output()
```
Might be a good idea to pass the output parameter as input, to avoid a copy.

zoq commented on src/mlpack/methods/neat/genome_impl.hpp:
```diff
@@ -0,0 +1,165 @@
+
+// Loads input.
+template <class ActivationFunction>
+void Genome<ActivationFunction>::Input(const arma::vec& input)
```
I guess we could remove the `Input` function here and just pass the input once we call the `Output` function. Let me know what you think.

favre49 replied:
Makes sense, I'll change the name to `Evaluate` then.

favre49 commented:
I've added the implementation for cyclic networks. I've used the same scheme for activation as in [SharpNEAT](https://github.com/colgreen/sharpneat/blob/master/src/SharpNeatLib/Phenomes/NeuralNets/FastCyclicNetwork/FastCyclicNetwork.cs) for now, but I'm not sure if its correct. Could you check if it's correct or suggest a better scheme for cyclic networks?

manish7294 commented on src/mlpack/methods/neat/cyclic_net.hpp:
```diff
@@ -0,0 +1,93 @@
+  /**
+   * Finds the output of the network from the given input.
+   */
+  arma::vec Evaluate(arma::vec input);
```
You should always consider passing by reference, otherwise you will be risking making an unncessary copy.

manish7294 commented on src/mlpack/methods/neat/cyclic_net_impl.hpp:
```diff
@@ -0,0 +1,83 @@
+
+  arma::vec output(outputNodeCount);
+  for (size_t i = 0; i < output.n_elem; i++)
+    output[i] = putputNodeValues[i + inputNodeCount + 1];
```
Typo!

manish7294 commented on src/mlpack/methods/neat/cyclic_net.hpp:
```diff
@@ -0,0 +1,93 @@
+  double bias;
+
+  //! The values stored by the nodes.
+  std::vector<double> outputNodeValues;
```
Would you like to use arma::vec instead for maintaining consistent with the codebase?

manish7294 commented on src/mlpack/methods/neat/cyclic_net_impl.hpp:
```diff
@@ -0,0 +1,83 @@
+      for (auto const &x : directedGraph[j])
+      {
+        double weight = x.second.getWeight();
+        inputNodeValues[x.first] += nodeValues[j] * weight;
```
I don't see any declaration of `nodeValues` throughout the cyclic net code. Are you yet to refine this impelementation as it looks kind of rough implementation?

favre49 replied:
Ah yes I've fixed some of these mistakes but am yet to push them. You're right, I'm yet to refine the cyclic net implementation. I plan to do so soon.

manish7294 commented on src/mlpack/methods/neat/acyclic_net.hpp:
```diff
@@ -0,0 +1,100 @@
+   * A digraph containing connection genes sorted by source ID, and then
+   * secondary sorted by target ID.
+   */
+  std::map<size_t, std::map<size_t, ConnectionGene>> directedGraph;
```
I am wondering, do we really need to store directedGraph here, I mean we can always pass the graph with evaluate right? Let me know your thoughts.

favre49 replied:
Good idea, I missed that. I'll see where else I can do something like this as well.

manish7294 commented on src/mlpack/methods/neat/acyclic_net.hpp:
```diff
@@ -0,0 +1,83 @@
+  /**
+   * Finds the output of the network from the given input.
+   */
+  arma::vec Evaluate(arma::vec& input,
```
You could actually even pass the ouput as a parameter to evaluate, instead of returning a copy. Your choice.

zoq replied:
Agreed, that is also what we do most of the time in other methods.

favre49 replied:
Sounds good, I'll do that.

manish7294 commented on src/mlpack/methods/neat/cyclic_net.hpp:
```diff
@@ -0,0 +1,80 @@
+   * Finds the output of the network from the given input.
+   */
+  arma::vec Evaluate(arma::vec& input,
+                     std::map<size_t, std::map<size_t, ConnectionGene>>& directedGraph);
```
Please wrap this line, to avoid inconsistency.

manish7294 commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,371 @@
+          class ActivationFunction,
+          class SelectionPolicy>
+bool NEAT<TaskType, ActivationFunction, SelectionPolicy>::
+    compareGenome(Genome<ActivationFunction> gen1,
```
As I mentioned earlier, always try using references, wherever applicable.

favre49 replied:
Ah yes, i sometimes forget. I'll make the changes, thanks!

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,92 @@
+  // Find the output.
+  arma::vec output(outputNodeCount);
+  for (size_t i = 0; i < output.n_elem; i++)
+    output[i] = nodeValues[i + inputNodeCount + 1];
```
To me it looks like we could sae an extra copy here if we use ``output`` for the assignments above instead of the ``std::vector``. Let me know what you think.

zoq replied:
We should use the activation function for the output as well.

zoq commented on src/mlpack/methods/neat/cyclic_net_impl.hpp:
```diff
@@ -0,0 +1,80 @@
+
+  arma::vec output(outputNodeCount);
+  for (size_t i = 0; i < output.n_elem; i++)
+    output[i] = outputNodeValues[i + inputNodeCount + 1];
```
See comment above.

zoq commented on src/mlpack/methods/neat/genome_impl.hpp:
```diff
@@ -0,0 +1,349 @@
+
+      // Change the weight of the gene in the directed graph as well.
+      directedGraph[connectionGeneList[i].source][connectionGeneList[i].target]
+          .setWeight() = connectionGeneList[i].getWeight();
```
To be consistent with the rest of the codebase, what about we use:

```c++
const double& Weight() const { return weight; }
double& Weight() { return weight; }
```

zoq commented on src/mlpack/methods/neat/selection_strategies/roulette_selection.hpp:
```diff
@@ -0,0 +1,54 @@
+        prob = 0;
+      }
+    }
+    return std::make_pair(i, j);
```
Do you think we could use ``arma::vec`` here as well, might be easier to work with the result.

zoq commented on src/mlpack/methods/neat/selection_strategies/roulette_selection.hpp:
```diff
@@ -0,0 +1,54 @@
+      prob += fitnesses[k] / totalFitness;
+      if (arma::randu<double>() < prob)
+      {
+        if (i == fitnesses.n_elem)
```
Maybe I missed something, but I think this is always true.

favre49 replied:
It ceases to be true the moment `i` is assigned an index for the first parent. After we do that, we can assign a parent to `j`.

zoq commented on src/mlpack/methods/neat/selection_strategies/roulette_selection.hpp:
```diff
@@ -0,0 +1,54 @@
+
+#include <mlpack/prereqs.hpp>
+
+class RouletteSelection
```
Can you write a test for each selection strategy.

favre49 replied:
So would that basically be an XOR test with all the different selection strategies?

zoq replied:
I thought we could write some specific tests for the strategies, pass in some defined fitness values and check that the output it correct, the XOR test would be an addition.

zoq commented:
In the last blog update, you said something about long build times, the build step should only rebuild changes, is that the case on your system as well? Perhaps it makes sense to comment the other methods locally to decrease the build time? Also I think I said this before but you can use multiple cores to build in parallel ``make -jx`` where x is the number of threads you like to use for the build.

zoq commented:
@favre49 Maybe you can help me here, is the last version you pushed the version that sometimes provides reasonable results?

favre49 commented:
@zoq Sorry for the late response, the newest commit should have no memory issues and also has connection deletion. The fitness is normally about 2.9, which is still subpar.

zoq commented:
Okay, have to step through the code.

zoq commented:
Have you locally implement a method to print a Genome? I can do this as well, just wanted to ask first.

favre49 commented:
There is the Parameters() method for that, but I also printed out the directed graph for easier-to-read connectivity data whenever I needed it. Haven't implemented it as a method though.

zoq commented:
Not sure I can see the printing of the directed graph, maybe it's not pushed? Ideally I can see every connection (enabled/disabled) and the corresponding weight, maybe the input and output as well. But I can get that from the ``Evaluate`` function as well. What I'm currently trying to see if NEAT is able to find a solution just by mutating the weights, which should be possible.

zoq commented:
If you don't have something like this, let me know, I can implement that function as well.

favre49 commented:
Ah, I have not implemented such a function. I just added print statements locally of whatever I needed.

zoq commented:
Perhaps something like:

```
Input: 1, 1
Output: 0.93
1 -> 2 (enabled): 0.212
1 -> 3 (enabled): 0.1234
1 -> 3 (enabled): -0.123
1 -> 3 (disabled): -0.123
```

this could work? Something that is easy to read.

zoq commented:
The first issue I can see is that the activation function isn't used for the last layer (output). Which often results in a output that blows up.

favre49 commented:
Makes sense, I'll make these changes.

favre49 commented:
Alright I've added a Print() function that does what you proposed.

> I can see that the output is really similar for every input around 0.8.

I'll look at this as well

zoq commented:
Is the printed topology already sorted? So for example:

```
   1.0000
   1.0000
   0.8948
0->3 (enabled) Weight: 0.95126
1->3 (enabled) Weight: 1.64916
2->3 (disabled) Weight: 2.70649
2->4 (enabled) Weight: 0.654655
4->3 (enabled) Weight: 0.839223
1->4 (enabled) Weight: 1.37963
```

will look like:

![IMG_0942](https://user-images.githubusercontent.com/4209744/59981061-d0deb980-95fe-11e9-860e-1634ca01e597.png)

and the output is gene 4?

favre49 commented:
No, the output node is always number 3 in this case. Since there is one bias node, two input nodes and one output node,  ID 0 is the bias node, ID 1 and 2 are the input nodes, and ID 3 is the output node. Anything with an ID more than that is a hidden node.

zoq commented:
Okay, thanks for the clarification, can you print the activation for the genes as well?

favre49 commented:
Yup, Print() does that as well now.

zoq commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,447 @@
+      size_t innovID2 = minGenome.connectionGeneList[j].InnovationID();
+      if (innovID2 < innovID1)
+      {
+        if (arma::randu<double>() < 0.5)
```
Shouldn't the threshold be configurable?

zoq commented on src/mlpack/methods/neat/selection_strategies/rank_selection.hpp:
```diff
@@ -0,0 +1,64 @@
+    // Choose first genome.
+    while (selection[0] == fitnesses.n_elem)
+    {
+      if (pos >= size)
```
I'm wondering if it makes performance wise to use the first index if we already iterated over. all.

zoq commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -0,0 +1,251 @@
+
+  double findError(arma::vec& input, Genome<HardSigmoidFunction>& genome)
+  {
+    arma::vec output = genome.Evaluate(input);
```
To match with the existing codebase, I think it makes sense to allow the user to pass an ``arma::mat``and internally iterate over the columns, instead of doing this outside the method. In this case we could pass the complete input data and use ``arma::accu(arma::pow(...))`` to get the error.

zoq commented on src/mlpack/methods/neat/gene_impl.hpp:
```diff
@@ -48,6 +48,17 @@ void ConnectionGene::Mutate(const double mutationSize)
 
+template <typename Archive>
+void ConnectionGene::serialize(Archive& ar,
+                                const unsigned int /* version */)
```
Pedantic style issue number one: the second parameter should align with the first one.

zoq commented on src/mlpack/methods/neat/gene_impl.hpp:
```diff
@@ -48,6 +48,17 @@ void ConnectionGene::Mutate(const double mutationSize)
+                                const unsigned int /* version */)
+{
+  ar & BOOST_SERIALIZATION_NVP(globalInnovationID);
+  ar & BOOST_SERIALIZATION_NVP(weight);
```
Should we save the weights only if we initialized the model?

zoq commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -197,79 +213,80 @@ class DPNVTask
+    else
+      fitness += timeStep;
+
+    std::cout <<" Fitness: " << fitness << std::endl;
```
In case we are going to provide such a function in the final version, we should go for ``Log::Debug``instead of ``std::cout``.

zoq commented:
Have you removed the Print functionality?

favre49 commented:
Oops must have removed it at some point. I'll add it back along with phased searching later tonight

zoq commented:
Okay, great, thanks.

favre49 commented:
I have added phased searching. NEAT now performs phased search if `complexityThreshold != 0`. I've done very basic testing, and it seems to be performing fine, except it might be unable to simplify further after a certain number of generations, I will have to investigate this further. If this is the case I might apply a relative complexity ceiling instead of an absolute one.

I also added the Print() function. On the non-Markovian double pole velocity experiment, I have only seen NEAT reach a fitness of about 450, though it should be able to get a fitness above a couple thousand.

zoq commented:
> I have added phased searching. NEAT now performs phased search if `complexityThreshold != 0`. I've done very basic testing, and it seems to be performing fine, except it might be unable to simplify further after a certain number of generations, I will have to investigate this further. If this is the case I might apply a relative complexity ceiling instead of an absolute one. I'll also implement a test more suited to this, like a many-input multiplexer or the OCR test talked about [here](https://sharpneat.sourceforge.io/phasedsearch.html).

Sounds great, will take a closer look at the code.

> I also added the Print() function. On the non-Markovian double pole velocity experiment, I have only seen NEAT reach a fitness of about 450, though it should be able to get a fitness above a couple thousand.

Great, thanks, should be helpful to track down the issue we see with the double pole velocity env.

manish7294 commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,540 @@
+    else
+      searchMode = 0;
+
+    std::cout << "Maximum fitness " << fitnesses.max() << std::endl;
```
Would be better to have them as Log::Info messages.

manish7294 commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,540 @@
+      speciesSize[i++]++;
+      delta--;
+      if(i >= numSpecies)
+        i = 0;
```
I guess you could do this a bit faster by mathematically calculating the number of complete repetion and the remaining offset. Though your choice. Same apply for the else if part.

manish7294 commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,540 @@
+    // This ensures that the best genome from every species will be saved.
+    if (numElite[i] == 0)
+      numElite[i] = 1;
+    if (numElite[i] > speciesList[i].size())
```
please put this in else if block, otherwise it would be an unccessary check.

zoq commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -0,0 +1,296 @@
+  double Evaluate(Genome<HardSigmoidFunction>& genome)
+  {
+    // Set the initial state.
+    typename EnvironmentType::State state = environment.InitialSample();
```
We want to evaluate if a model is all to solve a task using different start conditions, so this should move down into the for loop, so that we start with a new random position instead of using the previous one.

zoq commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -0,0 +1,296 @@
+        if (fitness > 10000)
+          break;
+      }
+      total += fitness;
```
We never reset the ``fitness`` value, so this is already the total fitness, I think we should reset the ``fitness`` parameter before each run, so that ``if (fitness > 10000)`` does work as well.

zoq replied:
Ahh, nevermind I can see you do reset the value.

zoq commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,601 @@
+    for (size_t j = 0; j < genomeList[i].connectionGeneList.size(); j++)
+    {
+      size_t innovID = genomeList[i].connectionGeneList[j].InnovationID();
+      data(innovID, i) = genomeList[i].connectionGeneList[j].Weight();
```
I think we should only account for enabled connections.

zoq commented on src/mlpack/tests/neat_test.cpp:
```diff
@@ -44,27 +44,27 @@ BOOST_AUTO_TEST_SUITE(NEATTest)
-  XORTask task;
-  NEAT<XORTask, HardSigmoidFunction, RankSelection> model(task, 2, 1, 100, 150, 10);
-  model.FinalFitness() = 3.8;
+// BOOST_AUTO_TEST_CASE(NEATXORTest)
```
I think as a quick example we should leave the XOR test as well.

favre49 replied:
Yup I'll leave it in, but do you think I should also write a short guide?

zoq commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -0,0 +1,155 @@
+class DPNVTask
+{
+ public:
+  DPNVTask(const MultiplePoleCart& env) : environment(env)
```
I'm wondering if the implementation of the task is wrong, I can see there is a test for the env: https://github.com/mlpack/mlpack/blob/c62088ad40c4f9a3839e23a5dbbd5ab55054c7ee/src/mlpack/tests/rl_components_test.cpp#L171 but we haven't really tested if e.g. QLearning is able to solve the env, maybe you tested it?

favre49 replied:
I think the learners we have right now aren't compatible with how the multiple pole cart class is structured. For instance, we use a matrix for the state data, and don't have a default constructor, for state and the environment. Maybe we should rewrite this class to allow for compatibility?

I've checked the equations again, and they are right according to the paper. I checked other implementations, and they are using RK4 along with these equations, which I don't know much about. I'll study about that and test that as well.

favre49 replied:
We could work around the issue of not having a default constructor for the state by making the multiple pole cart into an only double pole cart method. That way the API will be intact without having an odd exception environment. Moreover, I guess I haven't seen a triple or quadruple pole balancing test before (I made it general since I thought I could). Unless there's a better workaround?

zoq replied:
> We could work around the issue of not having a default constructor for the state by making the multiple pole cart into an only double pole cart method. That way the API will be intact without having an odd exception environment. 

That sounds like a reasonable solution to me.

> Moreover, I guess I haven't seen a triple or quadruple pole balancing test before (I made it general since I thought I could). Unless there's a better workaround?

Yeah, I really liked that.

> I was also thinking, unless it would make running the tests too lengthy, we should add unit tests for all these environments, where we check if one of our learners could solve them, so that we don't run into this problem or the one we had with the pendulum environment. It would make sure the API remains consistent. Let me know what you think.

Had the same idea, we could provide some initial parameter that are almost good enough, that way we could reduce the runtime for the tests.

favre49 replied:
I'll implement it locally for now, if NEAT works on it I'll open a PR to fix these issues.

zoq replied:
Sounds good, thanks!

zoq commented:
Id did some further testing, and it looks like there are some env's that try to maximise the reward and some to minimise the reward, an easy solution would be to use the absolute over the reward; Let me know what you think.

favre49 commented:
Yup, that makes sense to me.

zoq commented:
> Yup, that makes sense to me.

Should probably be an option, e.g. ``bool absoluteReward``, the default value (``true``?), just in case a user doesn't want to take the absolute value.

favre49 commented:
Actually, I think we should expect the user to ensure that fitness is mostly non-negative. Negative fitnesses don't really work with how species allocation is done, as it could make the species size negative (I handle this in the code by making negative species sizes zero, but it may not be a good idea to have fitnesses to be mostly negative). This seems like something for the user to take care of to me, what do you think?

zoq commented:
I see your point, so I guess it makes sense to check the output and to print out an error, if the fitness is negative. I tested a bunch of gym env's and some of them return a negative reward, so I had to make sure the reward is reasonable, in some cases taking the absolute reward works, but in some cases it doesn't. So, I agree might be a good idea to let the user handle it.

favre49 commented:
I fixed phased searching, and instead of giving it it's own test, I included it in the XOR task, and checked if the population's mean complexity was at most a unit above the current complexity ceiling as specified by phased searching . This is because it is possible that NEAT is in the process of simplification, in which case the complexity may be in excess of the ceiling (but it can't be too far above). Let me know what you think, if I should change the test or maybe split it into two, one for standalone XOR and one for Phased Searching

zoq commented:
Personally I would split the test in two, but looking at the current test time:  ``55/148 Test  #66: NEATTest .........................   Passed   24.56 sec`` I'm not sure if the current solution is preferable, we still have some other tests that aren't build around phased searching. Do you have the timings for the XOR test without and with phased searching? (To get a quick result I often use ```time bin/mlpack_test -t NEATTest/NEATXORTest``).

zoq commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,669 @@
+    // Set search mode.
+    if (meanComplexity > currentComplexityCeiling)
+    {
+      if (searchMode == 0)
```
I'm wondering if we should implement the search mode as a policy, so that we could add more methods in the future, without touching the NEAT implementation. Let me know what you think.

favre49 replied:
I had though of this, but I couldn't come up with a general policy implementation since I hadn't heard of any other search types. Moreover in this case it's achievable with just changing probabilities, but in other cases a search mode could require a different kind of mutation, new rules and measures (I'm just theorizing here, but phased searching involved keeping track of complexity metrics and adding a method to measure complexity) and we would have to change the NEAT implementation anyway. Is there a way I could come up with some kind of general policy representation?

zoq commented on doc/tutorials/neat/neat.txt:
```diff
@@ -0,0 +1,248 @@
+arma::vec input = arma::randn<arma::vec>(inputNodeCount);
+
+// Assuming a Genome object called genome has already been instantiated.
+arma::vec output = genome.Evaluate(input);
```
Do you think we should instantiate a Genome, that way a use could just copy past the lines from the example.

favre49 commented:
I checked, and phased searching has pretty much no effect on performance (probably makes it worse). This is to be expected though, since phased searching should only show it's usefulness in cases where there is the issue of genome bloat, not in easy tests like XOR. I only used this test to make sure phased test controlled complexity correctly.

zoq commented:
Okay, thanks for the clarification, I think the current test situation is fine.

zoq commented on src/mlpack/methods/neat/genome.hpp:
```diff
@@ -0,0 +1,319 @@
+ * @tparam ActivationFunction The activation function to be used.
+ */
+template <class ActivationFunction = ann::HardSigmoidFunction>
+class Genome
```
Since we use ``Genome<>`` a lot I think it would make sense from the style perspective, to use ``Genome`` instead, so we could write:

```
template <class ActivationFunction = ann::HardSigmoidFunction>
class GenomeType
{
};

template <class ActivationFunction = HardSigmoidFunction>
using Genome = GenomeType< ActivationFunction>;
```

Let me know what you think.

zoq commented on src/mlpack/methods/neat/neat.hpp:
```diff
@@ -0,0 +1,370 @@
+  /**
+   * Trains the model on the task and returns the best Genome.
+   */
+  Genome<ActivationFunction> Train();
```
To be consistent with the rest of the codebase I think it would make sense to pass a reference of the genome.

favre49 commented on src/mlpack/tests/neat_test_tools.hpp:
```diff
@@ -0,0 +1,149 @@
+/**
+ * @file neat_test_tools.hpp
```
I don't have an issue with that. The only con would be that separating them is more aesthetically pleasing, and easier to understand. I'll go with your judgement on this.

zoq replied:
Personally I find it easier to have them in the same file, as I don't have to open another file to search for the definition.

zoq commented on src/mlpack/tests/neat_test.cpp:
```diff
@@ -0,0 +1,77 @@
+ */
+BOOST_AUTO_TEST_CASE(NEATXORTest)
+{
+  arma::arma_rng::set_seed_random();
```
Can you remove the random seed here, for debugging it would be great to have the same random seed for each run.

zoq commented on src/mlpack/methods/neat/selection_strategies/tournament_selection.hpp:
```diff
@@ -0,0 +1,61 @@
+   * 
+   * @param fitnesses A sorted Armadillo vector of fitnesses in descending
+   *    order.
+   * @param contenderNum The number of contenders.
```
This is missing the ``selection`` vector.

zoq commented on src/mlpack/methods/neat/selection_strategies/tournament_selection.hpp:
```diff
@@ -0,0 +1,61 @@
+   * @param contenderNum The number of contenders.
+   * @param prob The probability of the fittest candidate being chosen.
+   */
+  static void Select(arma::vec& fitnesses,
```
Looks like the ``fitnesses`` parameter could be ``const``.

zoq commented on src/mlpack/methods/neat/selection_strategies/roulette_selection.hpp:
```diff
@@ -0,0 +1,53 @@
+   *    order.
+   * @param selection The selected indices.
+   */
+  static void Select(arma::vec& fitnesses, arma::uvec& selection)
```
See comment below about ``const``.

zoq commented on src/mlpack/methods/neat/selection_strategies/roulette_selection.hpp:
```diff
@@ -0,0 +1,53 @@
+        k--;
+      }
+      if (k == 0)
+        selection[i] = k;
```
I guess what could do is to set ``selection[i] = 0`` before the while loop, to avoid the extra check at the end.

zoq commented:
@mlpack-jenkins test this please

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,100 @@
+template <class ActivationFunction>
+void AcyclicNet<ActivationFunction>::Evaluate(const arma::vec& input,
+                                              arma::vec& output,
+                                              std::map<size_t, std::map<size_t,
```
Looks like this could be const.

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,100 @@
+                                              arma::vec& output,
+                                              std::map<size_t, std::map<size_t,
+                                                ConnectionGene>>& directedGraph,
+                                              std::vector<size_t>& nodeDepths)
```
Looks like this could be const as well.

favre49 replied:
Yup, I'll go through the codebase and add const wherever I can

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,100 @@
+  // Populate the layers.
+  for (size_t i = 0; i < nodeCount; i++)
+  {
+    while (layers.size() < nodeDepths[i] + 1)
```
Should be faster to call ``reserve`` once with the ``layers.size() - nodeDepths[i] + 1``.

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,100 @@
+    for (size_t j = 0; j < layers[i].size(); j++)
+    {
+      size_t nodeID = layers[i][j];
+      if (nodeID == 0)
```
Can you comment on the conditions here ``if (nodeID == 0)``, ``else if (nodeID <= inputNodeCount)`` and ``else``, I think it might help to understand the code.

zoq commented on src/mlpack/methods/neat/acyclic_net.hpp:
```diff
@@ -0,0 +1,84 @@
+ * into this class if the `isAcyclic` parameter is set to `true`. It is used
+ * to evaluate an input to the genome.
+ * 
+ * The steps of the decoding and activating are as follows:
```
This is really helpful, thanks!

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,106 @@
+            nodeValues[x.first] += bias * x.second.Weight();
+        }
+      }
+      // If it is an input node, we need not apply the activation function.
```
Do you think we should write "we don't need to apply the activation function"?

zoq commented on src/mlpack/tests/neat_test.cpp:
```diff
@@ -0,0 +1,210 @@
+{
+ public:
+  template<typename GenomeType>
+  double Evaluate(GenomeType& genome)
```
I think this can be const as well, since we don't change the state of the given genome.

favre49 replied:
Phased searching and cyclic evaluation both change Genome state

zoq replied:
So if I call ``Evaluate`` on a genome twice, same env settings, I could get different results in the testing phase?

zoq replied:
Let me know if I should clarify this point.

favre49 replied:
Oops sorry, I didn't get a notification for the comment.

Yes, since cyclic networks store their activation values. I could maybe add a reset function which would clear the old activation?

zoq replied:
So it's just used to store temporary results; I don't think this is a problem, the only issue I see is that from a user perspective this could mean that we actually change the object. But I think, it would have a negative effect on the runtime if we make a copy of the genome first. So unless you can see a simple solution, I think we should go with the non-const solution.

zoq commented on src/mlpack/methods/neat/acyclic_net_impl.hpp:
```diff
@@ -0,0 +1,106 @@
+      // If this is a bias node, we need to add bias to it's connections.
+      if (nodeID == 0)
+      {
+        for (auto const& x : directedGraph[nodeID])
```
We might have to change this to ``const auto const& x : directedGraph[nodeID]`` or I guess we could also use a standard for loop and use the index.

favre49 replied:
This would give an error, since there would be two const qualifiers

zoq replied:
Ohh, right I missed to remove the second const.

zoq commented on src/mlpack/methods/neat/genome_impl.hpp:
```diff
@@ -0,0 +1,523 @@
+void Genome<ActivationFunction>::DelConnMutation()
+{
+  size_t i = 0;
+  i = arma::randi<arma::uvec>(1, arma::distr_param(0,
```
You could use:

https://github.com/mlpack/mlpack/blob/12a50a055ba7f69340598329bd146ee37bec110f/src/mlpack/core/math/random.hpp#L106

zoq commented on src/mlpack/methods/neat/genome_impl.hpp:
```diff
@@ -0,0 +1,523 @@
+  for (size_t i = 0 ; i < connectionGeneList.size(); i++)
+  {
+    if (connectionGeneList[i].Enabled())
+      connCount++;
```
I'm wondering if it makes sense to build the connectivity at the same time as we add/disable the connection. Let me know what you think.

favre49 replied:
Do you mean update the nose depths? I tried this, but I couldn't think of a different way to do it.

zoq replied:
I was thinking we could update the connection count parameter once the genome is created, so if we can a new enabled connection we could increase the counter, if we disable one we could decrease the counter. That way we now at each time what the connection count is. Do you think that is feasible?

favre49 replied:
sounds good to me, I'll make that change

zoq replied:
Sounds good I think this could take some time.

zoq commented:
Let me restart the travis build, maybe this was just a temporary network higup.

zoq commented:
Do you think we could use the current model to solve a more complicated env something like Mario, I could set it up?

favre49 commented:
Sounds good, that would be a good showcase, if it works out.

zoq commented:
Okay, will set this up and let you know once there is something to show. This will has to run for a couple of iterations, I tested Q-Learning on Mario as well and if I remember right this took like 200k iterations.

zoq commented on src/mlpack/methods/neat/neat_impl.hpp:
```diff
@@ -0,0 +1,669 @@
+  for (size_t i = 0; i < connGeneList.size(); i++)
+  {
+    if (connGeneList[i].InnovationID() > maxInnovID)
+      maxInnovID = connGeneList[i].InnovationID;
```
This should be ``InnovationID()``.

zoq commented on src/mlpack/methods/neat/selection_strategies/tournament_selection.hpp:
```diff
@@ -0,0 +1,60 @@
+    {
+      arma::uvec contenders = arma::randi<arma::uvec>(contenderNum,
+          arma::distr_param(0, fitnesses.n_elem - 1));
+      arma::sort(contenders);
```
``arma::sort`` is not inplace.

zoq commented:
Looks like if I wanted to use https://github.com/Kautenja/gym-super-mario-bros I first have to implement titles support if I don't want to run NEAT with 184320 input nodes.

zoq commented:
Yes, let's keep this open, will update this with the numbers for the Mario env, soon.

abinezer commented:
hi mlpack! I am currently exploring neat genetic algos. Anything i could do to help a bit here?

favre49 commented:
Hey! This implementation is done, we wanted to test on super mario bros but got busy with other things. I'm hoping I get to merge this during the summer, if free. If testing the implementation is of interest to you, please feel free to do so :)

rcurtin commented:
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs. I *think* this one is relevant?)
